### DISCUSSION TITLE
Adding components

### DISCUSSION THREAD

**Discussion by ranjithshegde** [stevearc/overseer.nvim#7] [Q&A] [answered]
Firstly, 
Thanks for your wonderful work. This is such a comprehensive plugin! Finally something that can allow me to extend and customize thing I used to do with `vim-dispatch`

I have finally gotten around to defining and bundling templates (using `generator` where applicable).
Now starting to add/customize component for tasks.

What I am not able to figure out, if there is a way to pass any component while defining a template? for example something like
```lua
local pseudo_template = {
  name = "foo",
  desc = "something",
  components = {"here i pass it as list"} or function(_) end
}
```
If not, then what is the prefered way of attaching a component to a specific template or task?

**Comment by stevearc** [comment, marked as answer]
Templates have a `builder` function that returns a `overseer.TaskDefinition`. This is passed in directly to `overseer.new_task`. This could probably be documented better, but the type can be found here: https://github.com/stevearc/overseer.nvim/blob/57b1d3dcd5d3c27e86a48e1ba7efdb0e37dad380/lua/overseer/task.lua#L41-L49

So your template could look like this:
```lua
local my_template = {
  name = "My template",
  params = {},
  builder = function(params)
    return {
      cmd = "ls -l",
      components = {'default', 'my_custom_component'},
    }
  end
}
```

If you need to set the components on the _task_ based on configuration passed to your _template_, then you can add params to the template and use those in the builder to create the list of components.

**Reply by ranjithshegde** [reply]
Thank you for your response! That makes it fairly simple!

I seem to be completely useless at discovering these things, but how do I get `overseer` to recognize my custom_components?
I see that for templates there is `overseer.register_template()` 
In the Readme you say it has to be namespaced in overseer/component, does that mean I need to add them to the plugin folder?

for example, say I have some dummy component in a file under `.config/nvim/lua/mynamespace/tasks/component.lua`
```lua
return {
    desc = "print something",
    params = {},
    constructor = function(params)
        return {
            on_pre_start = function(self, task)
                vim.pretty_print { "some table" }
            end,
        }
    end,
}
```
How do I get overseer to see that file/component?

**Reply by stevearc** [reply]
Ah, this is good feedback! I should clarify the docs around this. You don't have to add anything directly to the plugin folder. For example, if overseer the plugin lives at `~/.local/share/nvim/site/pack/overseer/start/overseer.nvim`, you shouldn't need to directly touch that folder. All you have to do is make sure the file is added at the correct path (`overseer/component/mynamespace/mycomponent.lua`) somewhere in your runtimepath (`:help rtp`). In practice, the easiest will probably be inside your `.config/nvim`. So the full path to your custom component would be `~/.config/nvim/lua/overseer/component/mynamespace/mycomponent.lua`, and then you would add it to a task with `components = {"mynamespace.mycomponent", "default"}`.

**Reply by ranjithshegde** [reply]
Thank you! That helps
Plenty for me to play around with for a while!

Thanks for taking the time to explain these

**Reply by ranjithshegde** [reply]
I have one last question on this topic, if you dont mind.

Where (and when) do you pass the params defined in the components? Is it possible to set them in the template definition when setting `components = {"default", "myNamespace.some_component"}`? 

EDIT: I have tried 
`components = {"default", {"namespace.some_comp", param1 = "foo"}}`, 
`components = {"default", {"namespace.some_comp", {param1 = "foo"}}}`, 
`components = {"default", {"namespace.some_comp", params =  {param1 = "foo"}}}`, 

And they all throw errors

**Reply by stevearc** [reply]
That's very strange, because `components = {"default", {"namespace.some_comp", param1 = "foo"}}` is the correct syntax and should work. What errors are you seeing?

**Reply by ranjithshegde** [reply]
Apologies! I made a spelling mistake between template and component params!

It works as it should!
Thanks

**Reply by stevearc** [reply]
Ideally if a parameter name typo causes errors, the error message should clearly state what is going wrong. If it was cryptic or confusing, please file an issue and I'll make it more human-friendly

**Reply by ranjithshegde** [reply]
The error is readable. It states `componentName requires param myParams`

I had interpreted it to mean that param is not being passed because bad syntax, not that I typed it wrong.
